### DISCUSSION TITLE
How to get paramaterized URLs from react router for Sentry performance monitoring

### DISCUSSION THREAD

**Discussion by AbhiPrasad** [remix-run/react-router#8327] [Q&A]
Hey folks! 

I helped build the react router integrations for [Sentry's](https://sentry.io) frontend [performance monitoring](https://docs.sentry.io/product/performance/) product.

The [react router instrumentation](https://docs.sentry.io/platforms/javascript/guides/react/configuration/integrations/react-router/.
) listens to react router's `history` module to track pageloads and navigations for apps, all while displaying parameterized routes so that they grouped together for analysis (ex. `/teams/:teamid/user/:userid` -> `/teams/123/user/345`). So far we've created instrumentation for react router v3, v4, and v5.

Before we begin work on instrumentation for react router v6, we wanted to reach out and ask what y'all think is the best way for us to access parameterized routes with react router v6. We want it so that on history change, we can identify the url with the dynamic segments included. It's easy for us to get something like `/users/1234` through the hooks provided, but we want to grab `/users/:userid` (if it's configured by the user's routes).

From some initial playing around, it seems we can take advantage of `UNSAFE_RouteContext` to iterate through the matches, but we don't want to muck around with internals too much if we can avoid it (`UNSAFE` makes it sufficiently scary :P). We want to get the match that best defines the route that is being rendered.

We could also wrap individual `Route` components in a HOC and try to match based on what renders, but it seems hard to make a Route HOC due to this invariant check: https://github.com/remix-run/react-router/blob/263302da74a584aa5b825488ea6b22da69d62eb7/packages/react-router/index.tsx#L707-L712

Also thought about monkey patching `matchRoutes` - but obv want to avoid that if possible.

Opening as a discussion so we can brainstorm a little bit :)

**Comment by lukemorales** [comment]
@mjackson @ryanflorence @timdorr can we get an answer on this discussion?

**Comment by dcramer** [comment]
Bumping this as we’d love to support Remix. Suffice to say having a production grade developer experience is critical to tech adoption, so let us know how we can get this across the finish line.

**Comment by ryanflorence** [comment]
This is planned on [the roadmap](https://github.com/orgs/remix-run/projects/5/views/1) for both React Router and Remix.

We're working on an RFC. You can follow #9554 for updates.

**Reply by marisimon** [reply]
Hi Ryan! Any updates on this one? I know it’s an old item, but I couldn’t find #9554 in the roadmap.

**Reply by AbhiPrasad** [reply]
Hey @marisimon - you can take a look at https://github.com/getsentry/sentry-javascript/blob/develop/packages/react/src/reactrouterv6.tsx and https://github.com/getsentry/sentry-javascript/blob/develop/packages/react/src/reactrouterv7.tsx to see how we create paramaterized routes for react router and react router framework in the sentry sdks.

These aren't exposed by react router or remix in an easy way, so you'll have to add some custom code for now.
